File: string/458-findsubStringFromInfinite.py

'''
given an alphabet (can be lower or uppercase) string s and two integers i and j where i < j. 
Let’s say p is an infinite string of s repeating forever. Return the substring of p from indexes [i, j).

Constraints
1 ≤ n ≤ 100,000 where n is the length of s
0 ≤ i < j < 2 ** 31
Example 1
Input
s = “tiger”
i = 6
j = 8
Output
“ig”

Example 2
Input
s = “hi”
i = 2
j = 6
Output
“hihi”
'''

def findSubString(s,i,j):
    if not s or i<0 or j<i:
        return None
    
    n=len(s)
    k=j-i
    p=s*(k+1)
    return p[i:i+k]
    
def solve(s,i,j):
    n=len(s)
    ans=[]
    for id in range(i,j):
        ans.append(s[id%n])
    return "".join(ans)

s='tiger'

i = 6
j = 8
print(findSubString(s,i,j))
print(solve(s,i,j))

File: string/458-findsubStringFromInfinite.cpp
#include<iostream>
#include<string>

using STR=std::string;

STR findSubString(const STR& s,int i,int j)
{
    STR ans{};
    for(int idx=i;idx<j;++idx)
    {
        ans+=s[idx%s.size()];
    }
    return ans;
}
STR findSubString2(const STR& s,int i,int j)
{
    int n=s.size();
    int k=j-i;
    // STR p(k+1,s.c_str());
    STR p{};
    for(int c=0;c<k+1;++c)
        p+=s;
    return p.substr(i,k);
}

int main()
{
    STR s{"tiger"};//s{"hi"};//{"tiger"};
    int i=6,j=8;
    std::cout<<findSubString2(s,i,j);
}